core/src/OpenACC/Kokkos_OpenACC_Atomic.hpp

```diff
@@ -0,0 +1,1964 @@
+#define KOKKOS_OPENACC_ATOMIC_HPP
+
+#include <impl/Kokkos_Atomic_Memory_Order.hpp>
+#include <impl/Kokkos_Memory_Fence.hpp>
+#include <openacc.h>
+#include <stdio.h>
+#include <algorithm>
+
+//#define KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS_ERROR
+
+#if defined(KOKKOS_ENABLE_OPENACC_ATOMICS)
+namespace Kokkos {
+// OpenACC can do:
+// Operations: +, -, *, /, &, |, ^, <<, >>
+// Type: int, unsigned int, unsigned long long int
+// Variants:   
+// atomic_exchange/compare_exchange/assign/fetch_add/fetch_sub/fetch_mul/fetch_div/fetch_mod/fetch_max/fetch_min/fetch_and/fetch_or/fetch_xor/fetch_lshift/fetch_rshift/add_fetch/sub_fetch/mul_fetch/div_fetch/mod_fetch/max_fetch/min_fetch/and_fetch/or_fetch/xor_fetch/lshift_fetch/rshift_fetch/
+// _atomic_load/store
+
+// atomic_exchange -------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__ T atomic_exchange(
+    volatile T *const dest,
+    const T &val) {
+  T retval;
+  T *ptr = const_cast<T *>(dest);
+//  #pragma acc atomic capture
+  { retval = *ptr; *ptr = val; }
+  return retval;
+}
+
+#pragma acc routine seq
+template <>
+int atomic_exchange(
+    volatile int *const dest,
+    const int &val) {
+  int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicExch((int *)dest, val);
+#else
+  int *ptr = const_cast<int *>(dest);
+  #pragma acc atomic capture
+  { retval = *ptr; *ptr = val; }
+#endif
+  return retval;
+}
+
+#pragma acc routine seq
+template <>
+unsigned int atomic_exchange(
+    volatile unsigned int *const dest,
+    const unsigned int &val) {
+  unsigned int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicExch((unsigned int *)dest, val);
+#else
+  unsigned int *ptr = const_cast<unsigned int *>(dest);
+  #pragma acc atomic capture
+  { retval = *ptr; *ptr = val; }
+#endif
+  return retval;
+}
+
+#pragma acc routine seq
+template <>
+unsigned long long int atomic_exchange(
+    volatile unsigned long long int *const dest,
+    const unsigned long long int &val) {
+  unsigned long long int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicExch((unsigned long long int *)dest, val);
+#else
+  unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+  #pragma acc atomic capture
+  { retval = *ptr; *ptr = val; }
+#endif
+  return retval;
+}
+
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+#pragma acc routine seq
+template <>
+float atomic_exchange(
+    volatile float *const dest,
+    const float &val) {
+  float retval;
+  retval = atomicExch((float *)dest, val);
+  return retval;
+}
+#endif
+
+// atomic_compare_exchange -----------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__  T atomic_compare_exchange(
+    volatile T *const dest, const T &compare,
+    const T &val) {
+	if( acc_on_device(acc_device_not_host) ) {
+		#warning atomic_compare_exchange(T) not correctly implemented in OpenACC 
```

Comment:
It is an OpenACC problem; the current OpenACC standard does not support atomic_compare_exchange() operation.
I changed the code so that the atomic construct will abort for unsupported atomic operation types if running on a device. (On the host, it will sequentially execute the construct without atomicity.)